src/Data/Nat/Primality/Factorisation.agda

```diff
@@ -0,0 +1,211 @@
+--   encounter a factor, as we can then divide by that factor and continue from
+--   there without termination issues
+
+private
+  pattern 2+ n = suc (suc n)
+  pattern 2≤2+n = s≤s (s≤s z≤n)
+  pattern 1<2+n = 2≤2+n
+
+factorise : ∀ n → .{{NonZero n}} → Factorisation n
+factorise 1 = record
+  { factors = []
+  ; isFactorisation = refl
+  ; factorsPrime = []
+  }
+factorise (2+ n) = <-rec P factoriseRec (2 + n) {2} 2≤2+n (≤⇒≤‴ 2≤2+n) (2-rough-n (2 + n))
+  where
+
+  P : ℕ → Set
+  P n′ = ∀ {k} → 2 ≤ n′ → k ≤‴ n′ → k Rough n′ → Factorisation n′
+
+  factoriseRec : ∀ n → <-Rec P n → P n
+  factoriseRec (2+ n) rec (s≤s (s≤s n≤z)) ≤‴-refl k-rough-n = record
+    { factors = 2 + n ∷ []
+    ; isFactorisation = *-identityʳ (2 + n)
+    ; factorsPrime = k-rough-n ∷ []
+    }
+  factoriseRec (2+ n) rec {0} 2≤2+n (≤‴-step (≤‴-step k<n)) k-rough-n =
+    factoriseRec (2+ n) rec 2≤2+n k<n (2-rough-n (2+ n))
+  factoriseRec (2+ n) rec {1} 2≤2+n (≤‴-step k<n) k-rough-n =
+    factoriseRec (2+ n) rec 2≤2+n k<n (2-rough-n (2+ n))
+  factoriseRec (2+ n) rec {suc (suc k)} 2≤2+n (≤‴-step k<n) k-rough-n with 2 + k ∣? 2+ n
+  ... | no  k∤n = factoriseRec (2+ n) rec {3 + k} 2≤2+n k<n (extend-∤ k-rough-n k∤n)
+  ... | yes k∣n = record
+    { factors = 2 + k ∷ factors res
+    ; isFactorisation = prop
+    ; factorsPrime = roughn∧∣n⇒prime k-rough-n 2≤2+n k∣n ∷ factorsPrime res
+    }
+    where
+
+      q : ℕ
+      q = quotient k∣n
+
+      -- we know that k < n, so if q is 0 or 1 then q * k < n
+      2≤q : 2 ≤ q
+      2≤q = ≮⇒≥ (λ q<2 → contradiction (_∣_.equality k∣n) (>⇒≢ (prf (≤-pred q<2)))) where
+
+        prf : q ≤ 1 → q * 2+ k < 2 + n
+        prf q≤1 = begin-strict
+          q * 2+ k  ≤⟨ *-monoˡ-≤ (2 + k) q≤1 ⟩
+          2 + k + 0 ≡⟨ +-identityʳ (2 + k) ⟩
+          2 + k     <⟨ ≤‴⇒≤ k<n ⟩
+          2 + n     ∎ where open ≤-Reasoning
+
+      0<q : 0 < q
+      0<q = begin-strict
+        0 <⟨ s≤s z≤n ⟩
+        2 ≤⟨ 2≤q ⟩
+        q ∎ where open ≤-Reasoning
+
+      q<n : q < 2 + n
+      q<n = begin-strict
+        q           <⟨ m<m*n q (2 + k) ⦃ >-nonZero 0<q ⦄ 2≤2+n ⟩
+        q * (2 + k) ≡˘⟨ _∣_.equality k∣n ⟩
+        2 + n       ∎ where open ≤-Reasoning
+
+      q≮2+k : q ≮ 2 + k
+      q≮2+k q<k = k-rough-n 2≤q q<k (quotient∣n k∣n)
+
+      res : Factorisation q
+      res = rec q q<n {2 + k} 2≤q (≤⇒≤‴ (≮⇒≥ q≮2+k))
+          $ λ {d} d<k d-prime → k-rough-n d<k d-prime ∘ flip ∣-trans (quotient∣n k∣n)
+
+      prop : (2 + k) * product (factors res) ≡ 2 + n
+      prop = begin
+        (2 + k) * product (factors res)
+          ≡⟨ cong ((2 + k) *_) (isFactorisation res) ⟩
+        (2 + k) * q
+          ≡⟨ *-comm (2 + k) q ⟩
+        q * (2 + k)
+          ≡˘⟨ _∣_.equality k∣n ⟩
+        2 + n ∎ where open ≡-Reasoning
+
+------------------------------------------------------------------------
+-- Properties of a factorisation
+
+factorisation≥1 : ∀ {as} → All Prime as → product as ≥ 1
```

Comment:
This feels like it should live in `Data.Nat.Primality`? All this is saying is that the product of prime numbers is great than 1. It's not about a factorisation per se.